src/hist/plot.py

```diff
@@ -214,185 +262,408 @@ def plot2d_full(
+        )
+        return ret
+
+    return gauss
+
+
+def _fit_callable_to_hist(
+    model: Callable[[np.ndarray], np.ndarray],
+    _hist: hist.BaseHist,
+    likelihood: bool = False,
+) -> "Tuple[np.ndarray, np.ndarray, np.ndarray, Tuple[Tuple[float, ...], np.ndarray]]":
+    """
+    Fit a model, a callable function, to the histogram values.
+    """
+    variances = _hist.variances()
+    if variances is None:
+        raise RuntimeError(
+            "Cannot compute from a variance-less histogram, try a Weight storage"
+        )
+    hist_uncert = np.sqrt(variances)
+
+    # Infer best fit model parameters and covariance matrix
+    xdata = _hist.axes[0].centers
+    popt, pcov = _curve_fit_wrapper(
+        model, xdata, _hist.values(), hist_uncert, likelihood=likelihood
+    )
+    model_values = model(xdata, *popt)
+
+    if np.isfinite(pcov).all():
+        n_samples = 100
+        vopts = np.random.multivariate_normal(popt, pcov, n_samples)
+        sampled_ydata = np.vstack([model(xdata, *vopt).T for vopt in vopts])
+        model_uncert = np.nanstd(sampled_ydata, axis=0)
+    else:
+        model_uncert = np.zeros_like(hist_uncert)
+
+    return model_values, model_uncert, hist_uncert, (popt, pcov)
+
+
+def _plot_fit_result(
+    _hist: hist.BaseHist,
+    model_values: np.ndarray,
+    model_uncert: np.ndarray,
+    ax: matplotlib.axes.Axes,
+    eb_kwargs: Dict[str, Any],
+    fp_kwargs: Dict[str, Any],
+    ub_kwargs: Dict[str, Any],
+) -> FitResultArtists:
+    """
+    Plot fit of model to histogram data
+    """
+    x_values = _hist.axes[0].centers
+    variances = _hist.variances()
+    if variances is None:
+        raise RuntimeError(
+            "Cannot compute from a variance-less histogram, try a Weight storage"
+        )
+    hist_uncert = np.sqrt(variances)
+
+    errorbars = ax.errorbar(x_values, _hist.values(), hist_uncert, **eb_kwargs)
+
+    # Ensure zorder draws data points above model
+    line_zorder = errorbars[0].get_zorder() - 1
+    (line,) = ax.plot(x_values, model_values, **fp_kwargs, zorder=line_zorder)
+
+    # Uncertainty band for fitted function
+    ub_kwargs.setdefault("color", line.get_color())
+    if ub_kwargs["color"] == line.get_color():
+        ub_kwargs.setdefault("alpha", 0.3)
+    uncertainty_band = ax.fill_between(
+        x_values,
+        model_values - model_uncert,
+        model_values + model_uncert,
+        **ub_kwargs,
+    )
+
+    return FitResultArtists(line, errorbars, uncertainty_band)
+
+
+def plot_ratio(
+    _hist: hist.BaseHist,
+    ratio: np.ndarray,
+    ratio_uncert: np.ndarray,
+    ax: matplotlib.axes.Axes,
+    **kwargs: Any,
+) -> RatioArtists:
+    """
+    Plot a ratio plot on the given axes
+    """
+    x_values = _hist.axes[0].centers
+    left_edge = _hist.axes.edges[0][0]
+    right_edge = _hist.axes.edges[-1][-1]
+
+    # Set 0 and inf to nan to hide during plotting
+    ratio[ratio == 0] = np.nan
```

Comment:
> This is already supported. `hist_1.plot_ratio("normal")` will work

Are you sure? I think MyPy is right here. And in general, "will work" should be a synonym for "is tested". :P (though, I'd also trust mypy checking here, that's about 70% as effective as a test for something like this)

Reply:
> Are you sure?

Yes.

Reply:
Ahh, maybe this is never called by `hist.plot_ratio`, so it will truly always get an array. Maybe keeping the names distinct if one does not call the other would make it less confusing? `h.thing()` should call `hist.plot.thing()` or the names should not match.

Reply:
`plot_array_ratio`?

Reply:
I can parameterize the tests though to show this for all available `str` options for both `plot_pull` and `plot_ratio`

Reply:
Added test.

Reply:
> Ahh, maybe this is never called by `hist.plot_ratio` with identical arguments, so it will truly always get an array and it's my fault for adding `str` here. Maybe keeping the names distinct if one does not call the other would make it less confusing? `h.thing()` should call `hist.plot.thing(h)` or the names should not match.

Ah sorry, it seems that I'm missing some of your replies. I've renamed these in `plot.py` to `plot_ratio_array` and `plot_pull_array`.